### ISSUE TITLE
Write-through special device

### ISSUE THREAD

**Issue by Haravikk** [openzfs/zfs#15118]
### Describe the feature would like to see added to OpenZFS

The idea is to allow for `special` devices to be configured for a pool in a "write-through" mode such that any record assigned to a special vdev is also written to a non-special vdev as well, such that the special device does not require the same guarantee for redundancy as the rest of the pool.

This could be achieved using a `special_writethrough` pool property which, if enabled, will trigger the new behaviour. If this is enabled while there is an existing `special` vdev, any data within that vdev will be copied out to other devices until redundancy is achieved (the loss of the `special` devices no longer impacts redundancy). Once this process is complete, the `special` vdev can be freely removed. Disabling the feature would have no immediate impact, as any records already copied would remain copied, similar to existing records remaining where they are even if they qualify for the `special` device.

In the event that a special device is removed, and a new one later added, it would be possible for any records flagged as special to be copied onto the new device to "resilver" the special device. This should probably be an option when adding it, since it would require a search for qualifying records.

### How will this feature improve OpenZFS?

This will allow for the use of a `special` vdev with less (or no) redundancy compared to the rest of the pool, without the risk of data loss, while still accelerating reading of special records (metadata, small files etc.).

### Additional context

This feature is an alternative to #15051 (record size limit for ARC/L2ARC). While #15051 would be the easier to implement, this one is the more "correct" since it also accelerates smaller record reading but in a more predictable way since it can (with a suitable `special` device) guarantee that *all* such records are accelerated, rather than just whichever ones happen to remain in ARC/L2ARC long enough to be used.

**Comment by rincebrain** [comment]
Making a special mirror mode that would write new data out redundantly with a special case of copies=2-like behavior might be feasible, conceivably - rewriting it retroactively, I'd put my money on "not unless someone spends a fortune, and probably not then either".

**Comment by amotin** [comment]
Special vdev may be used to solve two problems -- speed and space efficiency.  From speed perspective proposed copies=2 logic may have sense, especially since a lot of metadata already use copies=2, and I don't think ZFS ever reads second copy if the first is OK.  It should not be a huge deal to write second copy to normal vdevs, but reconstruction of the lost special vdev is something not supported at this time.  From space efficiency though it makes no sense, since if main vdev is unable store small objects, like DRAID, then it just can't and redundant special vdev is the only solution to do it efficiently.

**Comment by Haravikk** [comment]
> Special vdev may be used to solve two problems -- speed and space efficiency. From speed perspective proposed copies=2 logic may have sense, especially since a lot of metadata already use copies=2, and I don't think ZFS ever reads second copy if the first is OK. It should not be a huge deal to write second copy to normal vdevs, but reconstruction of the lost special vdev is something not supported at this time. From space efficiency though it makes no sense, since if main vdev is unable store small objects, like DRAID, then it just can't and redundant special vdev is the only solution to do it efficiently.

Reconstruction isn't really a priority anyway, I'll tweak the original post to make that more clear. Currently when you add a `special` device you gain no benefit whatsoever until you start storing records the qualify for it, so there's no real need for the "write through" version to be any different.

"Reloading/preloading" a `special` device could probably be handled as a separate feature if necessary, since both cases would benefit from the ability to load up a special device from an existing pool that doesn't have one.

**Comment by GregorKopka** [comment]
Basically this asks for https://github.com/openzfs/zfs/issues/13460#issuecomment-1147375142

Maybe with the addition that all reads from eligible datasets resulting in cache misses are automatically written to this special vdev.

**Comment by Haravikk** [comment]
While there's similarity I'm not sure it's really the same thing; the advantage of `special` devices is that, sized correctly, they have a guarantee about the location of the special data and therefore predictable performance.

While "pinning" data in ARC is interesting, it has issues with potentially making the ARC less efficient (infrequently used pinned data preventing the ARC being used for more frequently used data) and how to specify the data to pin (and keep that updated in a useful way). I've submitted a similar but distinct proposal as #15051 to allow cache behaviour to be set as a record size; this is much simpler and would make it possible to prevent large records from filling up the ARC so more smaller records can be retained, and while this will help it doesn't provide the same guarantee that a correctly sized and configured `special` device can.

In general a `special` device is almost always superior to L2ARC, but it requiring redundancy and having no quick means to populate it means you can't just add a single `special` device and see benefits, but with "writethrough" behaviour and a way to preload it you could.

**Comment by GregorKopka** [comment]
I agree that special devices are superior to L2ARC in basically all aspects, except that **they can't be removed** and **they can't accelerate access to data already existing on-disk**.

The part you might have overlooked in the comment:
> But I guess it would make more sense to implement this feature as a new 'cache' vdev type (will call it C2 for this mental exercise) that, while being removeable at any time (like L2), employs random access writes by using space maps (similar to how normal data vdevs work), with a persistent lookup table (hosted on the C2-vdev itself) that translates from pool DVA into C2-LBA. Add a reader process that loads that lookup table on pool import to create C2 headers in the ARC (like persistent ARC). Then a writer process (that can be triggered eg. from changing a dataset property or a zpool subcommand) that will scan a specified dataset (or all that are marked for 'cache') to copy all eligble data from permanent pool vdevs onto the C2, so that new devices can be filled with data already existing in the pool (not only new writes). Last would be extending the read path in the same places L2 is hooked, a little patch to the L2 feeder logic to ignore data that is already in C2 (to avoid double-caching) and a SPA hook (I guess) to invalidate&free deleted data.

Also:
>While "pinning" data in ARC is interesting,...

I did not suggest anything like that.
